crypto/params.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+
+static const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->data_type != 0; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static int OSSL_PARAM_get_int_common(const OSSL_PARAM *p, const char *key,
+                                     void *val, size_t sz)
+{
+    const union {
+        long one;
+        char little;
+    } is_endian = { 1 };
+    int neg, sign;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    sign = p->data_type == OSSL_PARAM_INTEGER;
+    if (!sign && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)
+        return 0;
+    if (p->buffer_size == sz) {         /* Fast path */
+        memcpy(val, p->buffer, sz);
+    } else if (p->buffer_size < sz) {   /* Widening */
+        const size_t eb = sz - p->buffer_size;
+
+        if (is_endian.little) {
```

Comment:
I think this cleverness should be avoided, just use portable shift and or.

Reply:
That was actually cleverer than what I had in mind, which was something like this:

``` C
    const union {
        char digits[4];
        uint32_t number;
    } byteorder = { { 0x00, 0x01, 0x02, 0x03 } };
```

And then compare with `0x00010203`, `0x03020100`, `0x01000302` and whatever else to figure out the byte order (one of them is PDP-11, can you see which one?).

I actually think @paulidale's cleverness is much more to the point than most variants I've seen.  Of course, it demands that you know C...

Reply:
The union trick has been used by @dot-asm for crypto code.  I think it is needless tricky and premature optimization to use it for parameter-passing, especially when you can do things like always promote and then cast down:
```
case TYPE_SHORT: {
int itemp;
memcpy(item, param->value, sizeof (int));
*val = (short)itemp;
}
```

Reply:
So this is not really about the clever union trick, but rather about widening and narrowing being a bit over the top...  is that what you're trying to say?

Reply:
The union trick is needed only for the widening/narrowing optimization here. This code as currently proposed is a poster-child for needless optimization in my view.

Reply:
The union trick is from elsewhere in the OpenSSL code base, it isn't my creation.  Yes I remember the joy of Dec's own endian scheme.